File: backend/app/api/workflow_packages.py
# pyright: reportUnknownVariableType=false, reportUnknownMemberType=false, reportUnknownArgumentType=false, reportUnknownParameterType=false, reportMissingImports=false
from __future__ import annotations

from typing import Annotated, cast

from fastapi import APIRouter, Depends, Query, Request, Response, status

from app.api.dependencies import (
    get_workflow_package_runtime_input_registry_service,
    get_workflow_package_service,
)
from app.core.errors import validation_error
from app.schemas.workflow_package import (
    WorkflowPackageImportRequest,
    WorkflowPackageLaunchCreateRequest,
    WorkflowPackageLaunchCreateResponse,
    WorkflowPackageLaunchRead,
    WorkflowPackageListRead,
    WorkflowPackageManifestRead,
    WorkflowPackageManifestRequest,
    WorkflowPackagePreflightRequest,
    WorkflowPackageRead,
    WorkflowPackageRuntimeInputEntryRead,
    WorkflowPackageRuntimeInputPresetEntryCreateRequest,
    WorkflowPackageRuntimeInputPresetEntryUpdateRequest,
    WorkflowPackageRuntimeInputRegistryRead,
    WorkflowPackageSecretBindingListRead,
    WorkflowPackageSecretBindingRead,
    WorkflowPackageSecretBindingUpdateRequest,
    WorkflowPackageUpdateRequest,
    WorkflowPackageValidationRead,
)
from app.services.workflow_package_runtime_input_registry import (
    WorkflowPackageRuntimeInputRegistryService,
)
from app.services.workflow_package_service import WorkflowPackageService


def reject_removed_version_query(request: Request) -> None:
    if "version" not in request.query_params:
        return
    raise validation_error(
        "Workflow package request validation failed",
        [
            {
                "field": "version",
                "issue": "Workflow package version selection is no longer supported",
            }
        ],
    )


def reject_removed_status_query(request: Request) -> None:
    if "status" not in request.query_params:
        return
    raise validation_error(
        "Workflow package request validation failed",
        [
            {
                "field": "status",
                "issue": "Workflow package status filtering is no longer supported",
            }
        ],
    )


router = APIRouter(
    prefix="/workflow-packages",
    tags=["workflow-packages"],
    dependencies=[Depends(reject_removed_version_query)],
)


@router.get(
    "",
    response_model=WorkflowPackageListRead,
    dependencies=[Depends(reject_removed_status_query)],
)
def list_workflow_packages(
    service: Annotated[WorkflowPackageService, Depends(get_workflow_package_service)],
) -> WorkflowPackageListRead:
    return service.list_packages()


@router.post("", response_model=WorkflowPackageRead, status_code=status.HTTP_201_CREATED)
def create_workflow_package(
    payload: WorkflowPackageManifestRequest,
    service: Annotated[WorkflowPackageService, Depends(get_workflow_package_service)],
) -> WorkflowPackageRead:
    return service.create_package(payload)


@router.post("/validate-manifest", response_model=WorkflowPackageValidationRead)
def validate_workflow_package_manifest(
    payload: WorkflowPackageManifestRequest,
    service: Annotated[WorkflowPackageService, Depends(get_workflow_package_service)],
) -> WorkflowPackageValidationRead:
    return service.validate_manifest(payload)


@router.post("/import", response_model=WorkflowPackageRead, status_code=status.HTTP_201_CREATED)
def import_workflow_package(
    payload: WorkflowPackageImportRequest,
    service: Annotated[WorkflowPackageService, Depends(get_workflow_package_service)],
) -> WorkflowPackageRead:
    return service.import_package(payload)


@router.get("/{package_id}", response_model=WorkflowPackageRead)
def get_workflow_package(
    package_id: int,
    service: Annotated[WorkflowPackageService, Depends(get_workflow_package_service)],
) -> WorkflowPackageRead:
    return service.get_package(package_id)


@router.patch("/{package_id}", response_model=WorkflowPackageRead)
def update_workflow_package(
    package_id: int,
    payload: WorkflowPackageUpdateRequest,
    service: Annotated[WorkflowPackageService, Depends(get_workflow_package_service)],
) -> WorkflowPackageRead:
    return service.update_package(package_id, payload)


@router.delete(
    "/{package_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    response_model=None,
)
def delete_workflow_package(
    package_id: int,
    service: Annotated[WorkflowPackageService, Depends(get_workflow_package_service)],
) -> Response:
    service.delete_package(package_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get(
    "/{package_id}/secret-bindings",
    response_model=WorkflowPackageSecretBindingListRead,
)
def list_workflow_package_secret_bindings(
    package_id: int,
    service: Annotated[WorkflowPackageService, Depends(get_workflow_package_service)],
) -> WorkflowPackageSecretBindingListRead:
    return service.list_secret_bindings(package_id)


@router.put(
    "/{package_id}/secret-bindings/{key}",
    response_model=WorkflowPackageSecretBindingRead,
)
def upsert_workflow_package_secret_binding(
    package_id: int,
    key: str,
    payload: WorkflowPackageSecretBindingUpdateRequest,
    service: Annotated[WorkflowPackageService, Depends(get_workflow_package_service)],
) -> WorkflowPackageSecretBindingRead:
    return service.upsert_secret_binding(package_id, key, payload)


@router.delete(
    "/{package_id}/secret-bindings/{key}",
    status_code=status.HTTP_204_NO_CONTENT,
    response_model=None,
)
def delete_workflow_package_secret_binding(
    package_id: int,
    key: str,
    service: Annotated[WorkflowPackageService, Depends(get_workflow_package_service)],
) -> Response:
    service.delete_secret_binding(package_id, key)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get(
    "/{package_id}/runtime-input-registry",
    response_model=WorkflowPackageRuntimeInputRegistryRead,
)
def list_workflow_package_runtime_input_registry(
    package_id: int,
    workflow_key: Annotated[str, Query(alias="workflowKey")],
    service: Annotated[
        WorkflowPackageRuntimeInputRegistryService,
        Depends(get_workflow_package_runtime_input_registry_service),
    ],
) -> WorkflowPackageRuntimeInputRegistryRead:
    registry = service.list_registry(package_id, workflow_key)
    return WorkflowPackageRuntimeInputRegistryRead.model_validate(registry)


@router.post(
    "/{package_id}/runtime-input-registry/presets",
    response_model=WorkflowPackageRuntimeInputEntryRead,
    status_code=status.HTTP_201_CREATED,
)
def create_workflow_package_runtime_input_preset_entry(
    package_id: int,
    workflow_key: Annotated[str, Query(alias="workflowKey")],
    payload: WorkflowPackageRuntimeInputPresetEntryCreateRequest,
    service: Annotated[
        WorkflowPackageRuntimeInputRegistryService,
        Depends(get_workflow_package_runtime_input_registry_service),
    ],
) -> WorkflowPackageRuntimeInputEntryRead:
    entry = service.create_preset_entry(
        package_id,
        workflow_key,
        name=cast(str | None, payload.name),
        payload=payload.payload,
    )
    return WorkflowPackageRuntimeInputEntryRead.model_validate(entry)


@router.patch(
    "/{package_id}/runtime-input-registry/presets/{entry_id}",
    response_model=WorkflowPackageRuntimeInputEntryRead,
)
def update_workflow_package_runtime_input_preset_entry(
    package_id: int,
    entry_id: int,
    workflow_key: Annotated[str, Query(alias="workflowKey")],
    payload: WorkflowPackageRuntimeInputPresetEntryUpdateRequest,
    service: Annotated[
        WorkflowPackageRuntimeInputRegistryService,
        Depends(get_workflow_package_runtime_input_registry_service),
    ],
) -> WorkflowPackageRuntimeInputEntryRead:
    if "name" in payload.model_fields_set and "payload" in payload.model_fields_set:
        entry = service.update_preset_entry(
            package_id,
            workflow_key,
            entry_id,
            name=cast(str | None, payload.name),
            payload=payload.payload,
        )
    elif "name" in payload.model_fields_set:
        entry = service.update_preset_entry(
            package_id,
            workflow_key,
            entry_id,
            name=cast(str | None, payload.name),
        )
    elif "payload" in payload.model_fields_set:
        entry = service.update_preset_entry(
            package_id,
            workflow_key,
            entry_id,
            payload=payload.payload,
        )
    else:
        entry = service.update_preset_entry(package_id, workflow_key, entry_id)
    return WorkflowPackageRuntimeInputEntryRead.model_validate(entry)


@router.delete(
    "/{package_id}/runtime-input-registry/presets/{entry_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    response_model=None,
)
def delete_workflow_package_runtime_input_preset_entry(
    package_id: int,
    entry_id: int,
    workflow_key: Annotated[str, Query(alias="workflowKey")],
    service: Annotated[
        WorkflowPackageRuntimeInputRegistryService,
        Depends(get_workflow_package_runtime_input_registry_service),
    ],
) -> Response:
    service.delete_preset_entry(package_id, workflow_key, entry_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get("/{package_id}/manifest", response_model=WorkflowPackageManifestRead)
def get_workflow_package_manifest(
    package_id: int,
    service: Annotated[WorkflowPackageService, Depends(get_workflow_package_service)],
) -> WorkflowPackageManifestRead:
    return service.get_manifest(package_id)


@router.get("/{package_id}/export", response_model=None)
def export_workflow_package(
    package_id: int,
    service: Annotated[WorkflowPackageService, Depends(get_workflow_package_service)],
) -> Response:
    return service.export_package(package_id)


@router.post("/{package_id}/preflight", response_model=WorkflowPackageLaunchRead)
def preflight_workflow_package(
    package_id: int,
    payload: WorkflowPackagePreflightRequest,
    service: Annotated[WorkflowPackageService, Depends(get_workflow_package_service)],
) -> WorkflowPackageLaunchRead:
    return service.preflight_package(
        package_id,
        workflow_key=payload.workflow_key,
        parameters=payload.parameters,
    )


@router.get("/{package_id}/launch", response_model=WorkflowPackageLaunchRead)
def get_workflow_package_launch(
    package_id: int,
    service: Annotated[WorkflowPackageService, Depends(get_workflow_package_service)],
    workflow_key: Annotated[str | None, Query(alias="workflowKey")] = None,
) -> WorkflowPackageLaunchRead:
    return service.get_launch(package_id, workflow_key=workflow_key)


@router.post(
    "/{package_id}/launches",
    response_model=WorkflowPackageLaunchCreateResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_workflow_package_launch(
    package_id: int,
    payload: WorkflowPackageLaunchCreateRequest,
    service: Annotated[WorkflowPackageService, Depends(get_workflow_package_service)],
) -> WorkflowPackageLaunchCreateResponse:
    return service.create_launch(package_id, payload)

File: frontend/src/lib/api/workflow-packages.ts
import {
  buildPlatformApiUrl,
  requestPlatform,
  toPathSegment,
  toQueryRecord,
  type IdParam,
} from "../api-client";
import type {
  WorkflowPackageImportRequest,
  WorkflowPackageLaunchCreateRequest,
  WorkflowPackageLaunchCreateResponse,
  WorkflowPackageLaunchRead,
  WorkflowPackageListRead,
  WorkflowPackageManifestRead,
  WorkflowPackageManifestRequest,
  WorkflowPackagePreflightRequest,
  WorkflowPackageRead,
  WorkflowPackageRuntimeInputEntryRead,
  WorkflowPackageRuntimeInputPersonalEntryCreateRequest,
  WorkflowPackageRuntimeInputPersonalEntryUpdateRequest,
  WorkflowPackageRuntimeInputRegistryRead,
  WorkflowPackageSecretBindingListRead,
  WorkflowPackageSecretBindingRead,
  WorkflowPackageSecretBindingUpdateRequest,
  WorkflowPackageUpdateRequest,
  WorkflowPackageValidationRead,
} from "../types/workflow-package";

function workflowPackagePath(packageId: IdParam): string {
  return `/workflow-packages/${toPathSegment(packageId)}`;
}

function workflowPackageSecretBindingPath(packageId: IdParam, key: IdParam): string {
  return `${workflowPackagePath(packageId)}/secret-bindings/${toPathSegment(key)}`;
}

function workflowPackageRuntimeInputRegistryPath(packageId: IdParam): string {
  return `${workflowPackagePath(packageId)}/runtime-input-registry`;
}

function workflowPackageRuntimeInputPersonalEntryPath(
  packageId: IdParam,
  entryId: IdParam,
): string {
  return `${workflowPackageRuntimeInputRegistryPath(packageId)}/personal/${toPathSegment(entryId)}`;
}

type WorkflowPackageWorkflowOptions = {
  signal?: AbortSignal;
  workflowKey?: string | null;
};

type WorkflowPackageRuntimeInputRegistryOptions = {
  signal?: AbortSignal;
  workflowKey: string;
};

function workflowKeyQuery(options: WorkflowPackageWorkflowOptions = {}) {
  return toQueryRecord({
    workflowKey: options.workflowKey,
  });
}

export function listWorkflowPackages(signal?: AbortSignal): Promise<WorkflowPackageListRead> {
  return requestPlatform<WorkflowPackageListRead>("/workflow-packages", {
    signal,
  });
}

export function getWorkflowPackage(
  packageId: IdParam,
  signal?: AbortSignal,
): Promise<WorkflowPackageRead> {
  return requestPlatform<WorkflowPackageRead>(workflowPackagePath(packageId), { signal });
}

export function getWorkflowPackageManifest(
  packageId: IdParam,
  signal?: AbortSignal,
): Promise<WorkflowPackageManifestRead> {
  return requestPlatform<WorkflowPackageManifestRead>(`${workflowPackagePath(packageId)}/manifest`, {
    signal,
  });
}

export function validateWorkflowPackageManifest(
  payload: WorkflowPackageManifestRequest,
  signal?: AbortSignal,
): Promise<WorkflowPackageValidationRead> {
  return requestPlatform<WorkflowPackageValidationRead>("/workflow-packages/validate-manifest", {
    body: payload,
    method: "POST",
    signal,
  });
}

export function createWorkflowPackage(
  payload: WorkflowPackageManifestRequest,
  signal?: AbortSignal,
): Promise<WorkflowPackageRead> {
  return requestPlatform<WorkflowPackageRead>("/workflow-packages", {
    body: payload,
    method: "POST",
    signal,
  });
}

export function updateWorkflowPackage(
  packageId: IdParam,
  payload: WorkflowPackageUpdateRequest,
  signal?: AbortSignal,
): Promise<WorkflowPackageRead> {
  return requestPlatform<WorkflowPackageRead>(workflowPackagePath(packageId), {
    body: payload,
    method: "PATCH",
    signal,
  });
}

export function deleteWorkflowPackage(
  packageId: IdParam,
  signal?: AbortSignal,
): Promise<void> {
  return requestPlatform<void>(workflowPackagePath(packageId), {
    method: "DELETE",
    signal,
  });
}

export function listWorkflowPackageSecretBindings(
  packageId: IdParam,
  signal?: AbortSignal,
): Promise<WorkflowPackageSecretBindingListRead> {
  return requestPlatform<WorkflowPackageSecretBindingListRead>(`${workflowPackagePath(packageId)}/secret-bindings`, {
    signal,
  });
}

export function upsertWorkflowPackageSecretBinding(
  packageId: IdParam,
  key: IdParam,
  payload: WorkflowPackageSecretBindingUpdateRequest,
  signal?: AbortSignal,
): Promise<WorkflowPackageSecretBindingRead> {
  return requestPlatform<WorkflowPackageSecretBindingRead>(workflowPackageSecretBindingPath(packageId, key), {
    body: payload,
    method: "PUT",
    signal,
  });
}

export function deleteWorkflowPackageSecretBinding(
  packageId: IdParam,
  key: IdParam,
  signal?: AbortSignal,
): Promise<void> {
  return requestPlatform<void>(workflowPackageSecretBindingPath(packageId, key), {
    method: "DELETE",
    signal,
  });
}

export function getWorkflowPackageRuntimeInputRegistry(
  packageId: IdParam,
  options: WorkflowPackageRuntimeInputRegistryOptions,
): Promise<WorkflowPackageRuntimeInputRegistryRead> {
  return requestPlatform<WorkflowPackageRuntimeInputRegistryRead>(
    workflowPackageRuntimeInputRegistryPath(packageId),
    {
      query: workflowKeyQuery(options),
      signal: options.signal,
    },
  );
}

export function createWorkflowPackageRuntimeInputPersonalEntry(
  packageId: IdParam,
  payload: WorkflowPackageRuntimeInputPersonalEntryCreateRequest,
  options: WorkflowPackageRuntimeInputRegistryOptions,
): Promise<WorkflowPackageRuntimeInputEntryRead> {
  return requestPlatform<WorkflowPackageRuntimeInputEntryRead>(
    `${workflowPackageRuntimeInputRegistryPath(packageId)}/personal`,
    {
      body: payload,
      method: "POST",
      query: workflowKeyQuery(options),
      signal: options.signal,
    },
  );
}

export function updateWorkflowPackageRuntimeInputPersonalEntry(
  packageId: IdParam,
  entryId: IdParam,
  payload: WorkflowPackageRuntimeInputPersonalEntryUpdateRequest,
  options: WorkflowPackageRuntimeInputRegistryOptions,
): Promise<WorkflowPackageRuntimeInputEntryRead> {
  return requestPlatform<WorkflowPackageRuntimeInputEntryRead>(
    workflowPackageRuntimeInputPersonalEntryPath(packageId, entryId),
    {
      body: payload,
      method: "PATCH",
      query: workflowKeyQuery(options),
      signal: options.signal,
    },
  );
}

export function deleteWorkflowPackageRuntimeInputPersonalEntry(
  packageId: IdParam,
  entryId: IdParam,
  options: WorkflowPackageRuntimeInputRegistryOptions,
): Promise<void> {
  return requestPlatform<void>(workflowPackageRuntimeInputPersonalEntryPath(packageId, entryId), {
    method: "DELETE",
    query: workflowKeyQuery(options),
    signal: options.signal,
  });
}

export function importWorkflowPackage(
  payload: WorkflowPackageImportRequest,
  signal?: AbortSignal,
): Promise<WorkflowPackageRead> {
  return requestPlatform<WorkflowPackageRead>("/workflow-packages/import", {
    body: payload,
    method: "POST",
    signal,
  });
}

export function exportWorkflowPackageUrl(packageId: IdParam): string {
  return buildPlatformApiUrl(`${workflowPackagePath(packageId)}/export`);
}

export function preflightWorkflowPackage(
  packageId: IdParam,
  payload: WorkflowPackagePreflightRequest,
  signal?: AbortSignal,
): Promise<WorkflowPackageLaunchRead> {
  return requestPlatform<WorkflowPackageLaunchRead>(`${workflowPackagePath(packageId)}/preflight`, {
    body: payload,
    method: "POST",
    signal,
  });
}

export function getWorkflowPackageLaunch(
  packageId: IdParam,
  options: WorkflowPackageWorkflowOptions = {},
): Promise<WorkflowPackageLaunchRead> {
  return requestPlatform<WorkflowPackageLaunchRead>(`${workflowPackagePath(packageId)}/launch`, {
    query: workflowKeyQuery(options),
    signal: options.signal,
  });
}

export function createWorkflowPackageLaunch(
  packageId: IdParam,
  payload: WorkflowPackageLaunchCreateRequest,
  signal?: AbortSignal,
): Promise<WorkflowPackageLaunchCreateResponse> {
  return requestPlatform<WorkflowPackageLaunchCreateResponse>(`${workflowPackagePath(packageId)}/launches`, {
    body: payload,
    method: "POST",
    signal,
  });
}

export const workflowPackagesApi = {
  delete: deleteWorkflowPackage,
  deleteRuntimeInputPersonalEntry: deleteWorkflowPackageRuntimeInputPersonalEntry,
  deleteSecretBinding: deleteWorkflowPackageSecretBinding,
  create: createWorkflowPackage,
  createLaunch: createWorkflowPackageLaunch,
  createRuntimeInputPersonalEntry: createWorkflowPackageRuntimeInputPersonalEntry,
  exportUrl: exportWorkflowPackageUrl,
  get: getWorkflowPackage,
  getLaunch: getWorkflowPackageLaunch,
  getManifest: getWorkflowPackageManifest,
  getRuntimeInputRegistry: getWorkflowPackageRuntimeInputRegistry,
  import: importWorkflowPackage,
  list: listWorkflowPackages,
  listSecretBindings: listWorkflowPackageSecretBindings,
  preflight: preflightWorkflowPackage,
  update: updateWorkflowPackage,
  updateRuntimeInputPersonalEntry: updateWorkflowPackageRuntimeInputPersonalEntry,
  upsertSecretBinding: upsertWorkflowPackageSecretBinding,
  validateManifest: validateWorkflowPackageManifest,
} as const;
